envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoyListenersFactory.kt

```diff
@@ -0,0 +1,298 @@
+import io.envoyproxy.envoy.api.v2.listener.Filter
+import io.envoyproxy.envoy.api.v2.listener.FilterChain
+import io.envoyproxy.envoy.config.accesslog.v2.FileAccessLog
+import io.envoyproxy.envoy.config.filter.accesslog.v2.AccessLog
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpFilter
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.Rds
+import pl.allegro.tech.servicemesh.envoycontrol.groups.Group
+import com.google.protobuf.Any as ProtobufAny
+import io.envoyproxy.envoy.config.filter.http.header_to_metadata.v2.Config as HeaderToMetadataConfig
+
+typealias HttpFilterFactory = (node: Group) -> HttpFilter?
+
+@Suppress("MagicNumber")
+class EnvoyListenersFactory(
+    private val listenersFactoryProperties: ListenersFactoryProperties,
+    private val ingressFilters: List<HttpFilterFactory> = listOf(),
+    private val egressFilters: List<HttpFilterFactory> = listOf()
+) {
+    private val accessLogTimeFormat = stringValue(listenersFactoryProperties.httpFilters.accessLog.timeFormat)
+    private val accessLogMessageFormat = stringValue(listenersFactoryProperties.httpFilters.accessLog.messageFormat)
+    private val accessLogLevel = stringValue(listenersFactoryProperties.httpFilters.accessLog.level)
+    private val accessLogLogger = stringValue(listenersFactoryProperties.httpFilters.accessLog.logger)
+
+    companion object {
+        private const val egressRdsInitialFetchTimeout: Long = 20
+        private const val ingressRdsInitialFetchTimeout: Long = 30
+
+        val defaultServiceTagFilterRules = serviceTagFilterRules()
+        val defaultHeaderToMetadataConfig = headerToMetadataConfig(defaultServiceTagFilterRules)
+        val defaultHeaderToMetadataFilter = { _: Group -> headerToMetadataHttpFilter(defaultHeaderToMetadataConfig) }
+        val defaultEnvoyRouterHttpFilter = { _: Group -> envoyRouterHttpFilter() }
+        val defaultApiConfigSource: ApiConfigSource = apiConfigSource()
+        val defaultEgressFilters = listOf(defaultHeaderToMetadataFilter, defaultEnvoyRouterHttpFilter)
+        val defaultIngressFilters = listOf(defaultEnvoyRouterHttpFilter)
+
+        private fun envoyRouterHttpFilter(): HttpFilter = HttpFilter.newBuilder().setName("envoy.router").build()
+
+        fun apiConfigSource(): ApiConfigSource {
+            return ApiConfigSource.newBuilder()
+                    .setApiType(ApiConfigSource.ApiType.GRPC)
+                    .addGrpcServices(GrpcService.newBuilder()
+                            .setEnvoyGrpc(
+                                    GrpcService.EnvoyGrpc.newBuilder()
+                                            .setClusterName("envoy-control-xds")
+                            )
+                    ).build()
+        }
+
+        fun headerToMetadataHttpFilter(headerToMetadataConfig: HeaderToMetadataConfig.Builder): HttpFilter {
```

Comment:
I think `headerToMetadataHttpFilter` and `envoyRouterHttpFilter` could be simply a constants, instead of functions. This way they will be not re-calculated and allocated on every call.

Reply:
There is `defaultHeaderToMetadataFilter` and `defaultEnvoyRouterHttpFilter` which we initialize once. These functions are for convinience.

Reply:
But `defaultHeaderToMetadataFilter` and `defaultEnvoyRouterHttpFilter` are functions, so they call `headerToMetadataHttpFilter` every time.